crates/storage/nippy-jar/src/writer.rs

```diff
@@ -0,0 +1,341 @@
+        }
+
+        Ok(())
+    }
+
+    /// Appends a column to data file. `fn commit()` should be called to flush offsets and config to
+    /// disk.
+    pub fn append_column(
+        &mut self,
+        column: Option<ColumnResult<impl AsRef<[u8]>>>,
+    ) -> Result<(), NippyJarError> {
+        match column {
+            Some(Ok(value)) => {
+                if self.offsets.is_empty() {
+                    // Represents the offset of the soon to be appended data column
+                    self.offsets.push(self.data_file.stream_position()?);
+                }
+
+                self.write_column(value.as_ref())?;
+
+                // Last offset represents the size of the data file if no more data is to be
+                // appended. Otherwise, represents the offset of the next data item.
+                self.offsets.push(self.data_file.stream_position()?);
+            }
+            None => {
+                return Err(NippyJarError::UnexpectedMissingValue(
+                    self.jar.rows as u64,
+                    self.column as u64,
+                ))
+            }
+            Some(Err(err)) => return Err(err.into()),
+        }
+
+        Ok(())
+    }
+
+    /// Writes column to data file. If it's the last column of the row, call `finalize_row()`
+    fn write_column(&mut self, value: &[u8]) -> Result<(), NippyJarError> {
+        self.uncompressed_row_size += value.len();
+        if let Some(compression) = &self.jar.compressor {
+            let before = self.tmp_buf.len();
+            let len = compression.compress_to(value, &mut self.tmp_buf)?;
+            self.data_file.write_all(&self.tmp_buf[before..before + len])?;
+        } else {
+            self.data_file.write_all(value)?;
+        }
+
+        self.column += 1;
+
+        if self.jar.columns == self.column {
+            self.finalize_row();
+        }
+
+        Ok(())
+    }
+
+    /// Prunes rows from data and offsets file and updates its configuration on disk
+    pub fn prune_rows(&mut self, num_rows: usize) -> Result<(), NippyJarError> {
+        // Each column of a row is one offset
+        let num_offsets = num_rows * self.jar.columns;
+
+        // Calculate the number of offsets to prune from in-memory list
+        let offsets_prune_count = num_offsets.min(self.offsets.len().saturating_sub(1)); // last element is the expected size of the data file
+        let remaining_to_prune = num_offsets.saturating_sub(offsets_prune_count);
+
+        // Prune in-memory offsets if needed
+        if offsets_prune_count > 0 {
+            // Determine new length based on the offset to prune up to
+            let new_len = self.offsets[(self.offsets.len() - 1) - offsets_prune_count]; // last element is the expected size of the data file
+            self.offsets.truncate(self.offsets.len() - offsets_prune_count);
+
+            // Truncate the data file to the new length
+            self.data_file.set_len(new_len)?;
```

Comment:
if the program exits at this point, we have `data_file` truncated and the `offsets_file` with config still containing the old data.

Would consistency check work in this case? I see that it truncates offsets and data files according to the config (because it's written the last), and it works in case of program exit during append, but does it work during prune as well?